lib/services/label_expressions.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package services
+
+import (
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/gravitational/trace"
+	lru "github.com/hashicorp/golang-lru/v2"
+	"github.com/sirupsen/logrus"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/typical"
+)
+
+type labelExpression typical.Expression[labelExpressionEnv, bool]
+
+type labelExpressionEnv struct {
+	resourceLabelGetter LabelGetter
+	userTraits          map[string][]string
+}
+
+func parseLabelExpression(expr string) (labelExpression, error) {
+	if parsedExpr, ok := labelExpressionCache.Get(expr); ok {
+		return parsedExpr, nil
+	}
+	parsedExpr, err := labelExpressionParser.Parse(expr)
+	if err != nil {
+		return nil, trace.Wrap(err, "parsing label expression")
+	}
+	if evicted := labelExpressionCache.Add(expr, parsedExpr); evicted {
+		logrus.Warn("evicting entry from label expression cache")
```

Comment:
nit: Capitalize log messages.

```suggestion
		logrus.Warn("Evicting entry from label expression cache")
```

Also, is this worth a warning?